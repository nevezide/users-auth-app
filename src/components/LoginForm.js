import React from 'react';
import {useState} from 'react';

export default function LoginForm(props) {
  const [error, setError] = useState('');
  const [login, setLogin] = useState('');
  const [password, setPassword] = useState('');

  const onLogin = async () => {
    return fetch('http://localhost:3000/login', {
      method: 'POST',
      headers: {
        'content-type': 'application/json'
      },
      body: JSON.stringify({
        login,
        password
      })
    })
    .then((res) => {
      if (res.ok) {
        props.onLogged();
      } else {
        setError('Login / Password Incorrect')
      }
    })
    .catch(() => setError('An error occurs on login'));
  };

  return (
    <section className="form">
      <div className="error">{error}</div>
      <label htmlFor='login'>Login</label>
      <input type='text' id='login' name='login' placeholder='Votre email ou pseudo' onChange={(event) => setLogin(event.target.value)} value={login} />
      <label htmlFor='password'>Mot de passe</label>
      <input type='password' id='password' name='password' placeholder='Votre mot de passe' onChange={(event) => setPassword(event.target.value)} value={password} />
      <button onClick={onLogin}>S'identifier</button>
    </section>
  );
}
