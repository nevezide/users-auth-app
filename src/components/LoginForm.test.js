import React from 'react';
//import { toMatchDiffSnapshot } from 'snapshot-diff';
import {cleanup, fireEvent, waitFor, render} from '@testing-library/react';
import fetchMock from 'jest-fetch-mock';

import LoginForm from './LoginForm';

//expect.extend({ toMatchDiffSnapshot });
fetchMock.enableMocks();

beforeEach(() => {
  fetchMock.resetMocks();
});

// Remarque : l'exécution du nettoyage afterEach se fait automatiquement pour vous dans @testing-library/react@9.0.0 ou une version supérieure.
// Démonte et nettoie le DOM après que le test soit terminé.
afterEach(cleanup);

it('Show Login form', () => {
  const isLogged = false;
  const component = render(
    <LoginForm isLogged={isLogged} />,
  );

  expect(component.container).toMatchSnapshot()
});

it('Login succeed', async () => {
  const onLogged = jest.fn();
  const { getByRole, getByLabelText, asFragment } = render(
    <LoginForm onLogged={onLogged} />,
  );
  fetchMock.mockResponseOnce((req) => {
    try {
      expect(req.url).toBe('http://localhost:3000/login');
      expect(req.method).toBe('POST');
      expect(req.headers.get('Content-Type')).toBe('application/json');

      expect(JSON.parse(req.body)).toStrictEqual({ login: 'nevezide', password: 'nevezide' });

      return Promise.resolve(JSON.stringify({ data: 'Hello World' }));
    } catch(error) {
      console.error(error);
      throw error;
    }
  });
  const firstRender = asFragment();

  fireEvent.change(getByLabelText(/login/i), {target: {value: 'nevezide'}})
  fireEvent.change(getByLabelText(/mot de passe/i), {target: {value: 'nevezide'}})
  fireEvent.click(getByRole('button'));

  await waitFor(() => expect(onLogged).toHaveBeenCalledTimes(1));

  const lastFragment = asFragment();

  //expect(firstRender).toMatchDiffSnapshot(lastFragment)

  // Nettoyage des mocks
  fetchMock.mockReset();
});
