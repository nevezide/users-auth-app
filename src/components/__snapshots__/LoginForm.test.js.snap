// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login succeed 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -11,11 +11,11 @@
      <input
        id=\\"login\\"
        name=\\"login\\"
        placeholder=\\"Votre email ou pseudo\\"
        type=\\"text\\"
-       value=\\"\\"
+       value=\\"nevezide\\"
      />
      <label
        for=\\"password\\"
      >
        Mot de passe
@@ -23,11 +23,11 @@
      <input
        id=\\"password\\"
        name=\\"password\\"
        placeholder=\\"Votre mot de passe\\"
        type=\\"password\\"
-       value=\\"\\"
+       value=\\"nevezide\\"
      />
      <button>
        S'identifier
      </button>
    </form>"
`;

exports[`Show Login form 1`] = `
<div>
  <section
    class="form"
  >
    <div
      class="error"
    />
    <label
      for="login"
    >
      Login
    </label>
    <input
      id="login"
      name="login"
      placeholder="Votre email ou pseudo"
      type="text"
      value=""
    />
    <label
      for="password"
    >
      Mot de passe
    </label>
    <input
      id="password"
      name="password"
      placeholder="Votre mot de passe"
      type="password"
      value=""
    />
    <button>
      S'identifier
    </button>
  </section>
</div>
`;
